services/teamcity/teamcity-build.service.js

```diff
@@ -1,124 +1,115 @@
 'use strict'
 
-const LegacyService = require('../legacy-service')
-const { makeBadgeData: getBadgeData } = require('../../lib/badge-data')
+const Joi = require('joi')
+const TeamCityBase = require('./teamcity-base')
 
-// This legacy service should be rewritten to use e.g. BaseJsonService.
-//
-// Tips for rewriting:
-// https://github.com/badges/shields/blob/master/doc/rewriting-services.md
-//
-// Do not base new services on this code.
-function teamcityBadge(
-  url,
-  buildId,
-  advanced,
-  format,
-  data,
-  sendBadge,
-  request
-) {
-  const apiUrl = `${url}/app/rest/builds/buildType:(id:${buildId})?guest=1`
-  const badgeData = getBadgeData('build', data)
-  request(
-    apiUrl,
-    { headers: { Accept: 'application/json' } },
-    (err, res, buffer) => {
-      if (err != null) {
-        badgeData.text[1] = 'inaccessible'
-        sendBadge(format, badgeData)
-        return
-      }
-      try {
-        const data = JSON.parse(buffer)
-        if (advanced)
-          badgeData.text[1] = (
-            data.statusText ||
-            data.status ||
-            ''
-          ).toLowerCase()
-        else badgeData.text[1] = (data.status || '').toLowerCase()
-        if (data.status === 'SUCCESS') {
-          badgeData.colorscheme = 'brightgreen'
-          badgeData.text[1] = 'passing'
-        } else {
-          badgeData.colorscheme = 'red'
-        }
-        sendBadge(format, badgeData)
-      } catch (e) {
-        badgeData.text[1] = 'invalid'
-        sendBadge(format, badgeData)
-      }
+const buildStatusRegex = /SUCCESS|FAILURE|ERROR/
+const buildStatusTextRegex = /^Success|Failure|Error|Tests( failed: \d+( \(\d+ new\))?)?(,)?( passed: \d+)?(,)?( ignored: \d+)?(,)?( muted: \d+)?$/
+const buildStatusSchema = Joi.object({
+  status: Joi.string()
+    .regex(buildStatusRegex)
+    .required(),
+  statusText: Joi.string()
+    .regex(buildStatusTextRegex)
+    .required(),
+}).required()
+
+module.exports = class TeamCityBuild extends TeamCityBase {
+  static render({ status, statusText, useVerbose }) {
+    let color = 'red'
+    let message = status.toLowerCase()
+
+    if (useVerbose) {
+      message = (statusText || status).toLowerCase()
     }
-  )
-}
 
-module.exports = class TeamcityBuild extends LegacyService {
+    if (status === 'SUCCESS') {
+      color = 'brightgreen'
+      message = 'passing'
+    }
```

Comment:
I couldn't understand this at first. This way of writing it seems clearer:

```js
if (status === 'SUCCESS') {
  return {
    message: 'passing',
    color: 'brightgreen',
  }
} else if (statusText && useVerbose) {
  return {
    message: statusText.toLowerCase(),
    color: 'red'
  }
} else {
  return {
    message: status.toLowerCase(),
    color = 'red',
  }
}
```

Reply:
Agreed!